psydac/api/tests/test_api_2d_compatible_spaces.py

```diff
@@ -0,0 +1,175 @@
+
+    x = domain.coordinates
+
+    u,v = [VectorTestFunction(V1, name=i) for i in ['u', 'v']]
+    p,q = [ScalarTestFunction(V2, name=i) for i in ['p', 'q']]
+
+    a  = BilinearForm(((u,p),(v,q)),inner(grad(u),grad(v)) + div(u)*q - p*div(v) )
+    l  = LinearForm((v,q), f1*v[0]+f2*v[1]+q)
+    
+    bc = EssentialBC(u, 0, domain.boundary)
+    equation = find([u,p], forall=[v,q], lhs=a((u,p),(v,q)), rhs=l(v,q), bc=bc)
+
+    # ... create the computational domain from a topological domain
+    domain_h = discretize(domain, ncells=ncells)
+    # ... discrete spaces
+    V1h = discretize(V1, domain_h, degree=degree)
+    V2h = discretize(V2, domain_h, degree=degree)
+    Xh  = discretize(X , domain_h, degree=degree)
+    
+    s11,s12 = V1h.spaces[0].vector_space.starts
+    e11,e12 = V1h.spaces[0].vector_space.ends
+    s21,s22 = V1h.spaces[1].vector_space.starts
+    e21,e22 = V1h.spaces[1].vector_space.ends
+    s31,s32 = V2h.vector_space.starts
+    e31,e32 = V2h.vector_space.ends
+    
+    # ... dsicretize the equation using Dirichlet bc
+    ah = discretize(equation, domain_h, [Xh, Xh], symbolic_space=[X, X])
+
+    # ...
+    M = ah.assemble()
+    
+    M     = ah.linear_system.lhs
+    M[0,0][0,:,0,0] = 1.
+    M[0,0][:,0,0,0] = 1.
+    M[0,0][e11,:,0,0] = 1.
+    M[0,0][:,e12,0,0] = 1.
+    M[1,1][0,:,0,0] = 1.
+    M[1,1][:,0,0,0] = 1.
+    M[1,1][e21,:,0,0] = 1.
+    M[1,1][:,e22,0,0] = 1.
+    M = M.toarray()
+    
+    rhs   = ah.linear_system.rhs.toarray()
+
+    M_1   = np.zeros((len(rhs)+1,len(rhs)+1))
+    rhs_1 = np.zeros(len(rhs)+1)
+    v2h_s = (e11+1)*(e12+1)+(e21+1)*(e22+1)
+
+    M_1[:-1,:-1] = M
+    M_1[-1,v2h_s:-1]  = rhs[v2h_s:]
+    M_1[v2h_s:-1,-1]  = rhs[v2h_s:]
+    rhs_1[:v2h_s]   = rhs[:v2h_s] 
+
+    M_inv = linalg.inv(M_1)
+    sol = M_inv.dot(rhs_1)
+    
+    phi2   = FemField(V2h)
+    phi1_0 = FemField(V1h.spaces[0])
+    phi1_1 = FemField(V1h.spaces[1])
+    
+    phi1_0.coeffs[s11:e11+1, s12:e12+1] = sol[:(e11+1)*(e12+1)].reshape((e11+1-s11, e12+1-s12))
+    phi1_1.coeffs[s21:e21+1, s22:e22+1] = sol[(e11+1)*(e12+1):v2h_s].reshape((e21+1-s21, e22+1-s22))
+    phi2.coeffs[s31:e31+1, s32:e32+1]   = sol[v2h_s:-1].reshape((e31-s31+1, e32-s32+1))
+    
+    return sol
+
+###############################################################################
+#            SERIAL TESTS
+###############################################################################
+
+#==============================================================================
+def test_api_system_2_2d_dir_1():
+    from sympy.abc import x,y
+    from sympy import cos
+    
+    f1 = -x**2*(x - 1)**2*(24*y - 12) - 4*y*(x**2 + 4*x*(x - 1) + (x - 1)**2)*(2*y**2 - 3*y + 1) - 2*pi*cos(2*pi*x)
+    f2 = 4*x*(2*x**2 - 3*x + 1)*(y**2 + 4*y*(y - 1) + (y - 1)**2) + y**2*(24*x - 12)*(y - 1)**2 + 2*pi*cos(2*pi*y)
+    u1 = x**2*(-x + 1)**2*(4*y**3 - 6*y**2 + 2*y)
+    u2 =-y**2*(-y + 1)**2*(4*x**3 - 6*x**2 + 2*x)
+    p  = sin(2*pi*x) - sin(2*pi*y)
```

Comment:
Given that we have an analytical solution for `u1`, `u2` and `p`, it should be possible to assemble and compute the norm of the numerical solution.

Then you should (manually) refine the mesh and make sure that the scheme approaches the exact solution with the correct order of convergence.

Finally, once you are convinced that the unit test is working correctly, you should add an error tolerance for a given mesh and an `assert` statement. (See file **test_api_2d_system.py**)

Reply:
the tests are not complete yet ,we have the convergence but not the order, @ahmed is gonna take a look at that later ,so for the moment the tests are here to confirm that the mechanism works but i will change them later

Reply:
I see, but I think that in any case you should assemble and compute the error norms.

Reply:
okay i will add that